Adds jupyter-run to jupyter_client

dsblank opened the pull request.

takluyver commented on jupyter_client/runapp.py:
```diff
@@ -0,0 +1,146 @@
+                stream = getattr(sys, content['name'])
+                stream.write(content['text'])
+            elif msg_type in ('display_data', 'execute_result', 'error'):
+                if msg_type == 'error':
```
This nesting is a bit odd - why not have `elif msg_type == 'error'` broken out separately before display_data and execute_result.

dsblank replied:
I think this is cut and paste from Min's link:

https://github.com/dask/distributed/pull/370/files#diff-b7f744240c50427e50e9c844a13a00fdR43

so @minrk can comment on that.

takluyver commented on jupyter_client/runapp.py:
```diff
@@ -0,0 +1,146 @@
+        super(RunApp, self).start()
+        for filename in self.filenames_to_run:
+            self.log.debug("jupyter run: executing `%s`" % filename)
+            cell = open(filename).read()
```
As is good practice in general, we should ensure the file is closed deterministically:

```python
with open(filename) as f:
    cell = f.read()
```

dsblank replied:
Good catch; thanks. Pushed fix.

takluyver commented on jupyter_client/runapp.py:
```diff
@@ -0,0 +1,146 @@
+            cell = open(filename).read()
+            self.run_cell(cell)
+
+    def run_cell(self, cell):
```
We already have several `run_cell` methods, so I'd rather call this something more precise. Maybe `execute_printing_output`?

takluyver commented:
If I do `jupyter run foo.py`, and it errors, I certainly think the process should exit with code 1, so that e.g. CI knows something has gone wrong.

If I do `jupyter run foo.py bar.py` and there's an error running `foo.py`, IMO it should stop without running `bar.py`.

dsblank commented:
Agreed regarding stopping-on-error. I added two checks for error conditions, as I found some kernels had not yet used the error message type, so printing on stderr is also detected as an error code here.

dsblank commented:
Mentioning https://github.com/jupyter/jupyter_client/issues/183 as I forgot before. This should take care of that issue.

takluyver commented:
Sorry, I don't think printing to stderr should count as an error - it's common to write logging and things to stderr even when there's no exception condition.

We should, however, get the `execute_reply` message from the shell channel and check for `'status': 'error'` or `'status': 'aborted'` on that.

dsblank commented:
@takluyver I agree about counting as an error. However, there some kernels that need to be update to be able to handle this. I'll add "aborted" too.

takluyver commented on jupyter_client/runapp.py:
```diff
@@ -0,0 +1,157 @@
+                    return_code = 1
+                else:
+                    sys.stdout.write(content['data'].get('text/plain', ''))
+            elif msg_type == 'aborted':
```
This is not a message type. We need to get a message from the shell channel with `self.kernel_client.get_shell_msg`, check that it's an `execute_reply` message, and check its 'status' field.

http://jupyter-client.readthedocs.io/en/latest/messaging.html#execution-results

takluyver commented on jupyter_client/runapp.py:
```diff
@@ -0,0 +1,157 @@
+                stream = getattr(sys, content['name'])
+                stream.write(content['text'])
+                # TODO: remove this when all kernels use error msg_type:
+                if content['name'] == "stderr":
```
No, really, we shouldn't do this. It breaks kernels that do signal errors correctly to make a vague affordance to kernels that don't.

I suspect that when we correctly check the `execute_reply` message in addition to this, we'll find that most kernels do work correctly.

minrk commented:
Exit code should only consider the `status` field of the `execute_reply`. Outputs shouldn't be relevant.

dsblank commented:
Removed the break condition when printing to stderr, and removed the "aborted" msg_type handler. Somehow, this needs to poll handle iopub and shell, and print output from both. But I don't think @minrk 's dask link was doing this, right?

dsblank commented:
Oh, I think I see my issue: not all kernels publish the execute_reply also on the iopub channel. If they do, I think that this will work. What would probably also make this more clear is:

* documenting what is done with the error message parts (evalue, and ename; that traceback is a list of lines ending with a newline, etc)
* simple_kernel should show how to handle errors

takluyver commented:
No kernels should publish `execute_reply` on the iopub channel.

When the kernel gets an `execute_request` message on the shell channel, it sends 0 to N output messages on iopub, which may include messages with msg_type `error` (in addition to stream, display_data, execute_result). Then it sends precisely one `execute_reply` message on the shell channel, which indicates that the requested execution is done, and includes a status field (ok/error/aborted).

So the error information can be duplicated: in an `error` message on iopub, and the `execute_reply` message on shell. This is certainly somewhat confusing, and we hope to simplify it in a future version of the message spec, but that's what we have for now.

There should be no need to poll shell or print output from it. We can poll iopub until we get the status:idle message telling us that we've seen all outputs. And then we can fetch the execute_reply message from shell.

dsblank commented:
@takluyver Ok, thanks. That is confusing, and isn't reflected in the code that @minrk linked to, that I can see. I don't see how the code would keep the error message from printing out twice, if it can come in two forms (one on iopub and one on shell).

I think I'll wait for someone to write the ultimate function for handling this. Then I'll call it from this code.

minrk commented:
@dsblank thanks! I opened #185 adding BlockingKernelClient.run, which should be the implementation of the messaging. After that gets in, you should be able to simplify this one with:

```python
reply = self.kernel_client.run(code)
exit_code = 0 if reply['content']['status'] == 'ok' else 1
```

dsblank commented:
Updated PR to use reply=True. Very nice!

dsblank commented:
I think that this can be tagged 4.5 like https://github.com/jupyter/jupyter_client/pull/185

rgbkrk commented:
Neat.

minrk commented:
@dsblank sorry for the delay on this one. Thanks!

Marked as 5.0, since that's what the next release will be.
